numba/core/codegen.py

```diff
@@ -659,9 +660,13 @@ def _optimize_functions(self, ll_module):
             # module-level optimization.
             for func in ll_module.functions:
                 k = f"Function passes on {func.name!r}"
+                remarks_dir = os.environ.get("NUMBA_LLVM_OPT_REMARKS_DIR", None)
+                remarks_file = os.path.join(remarks_dir,
+                                            "%s-remarks.yaml" % func.name
+                                            ) if remarks_dir else None
```

Comment:
This needs to be moved to https://github.com/numba/numba/blob/203d6548a6f84e51f7346c47da271a47799f187d/numba/core/codegen.py#L687. It is where all the aggressive optimization take place.

Something like:

```diff
diff --git a/numba/core/codegen.py b/numba/core/codegen.py
index 472c7aede..17a1342ff 100644
--- a/numba/core/codegen.py
+++ b/numba/core/codegen.py
@@ -660,13 +660,9 @@ class CPUCodeLibrary(CodeLibrary):
             # module-level optimization.
             for func in ll_module.functions:
                 k = f"Function passes on {func.name!r}"
-                remarks_dir = os.environ.get("NUMBA_LLVM_OPT_REMARKS_DIR", None)
-                remarks_file = os.path.join(remarks_dir,
-                                            "%s-remarks.yaml" % func.name
-                                            ) if remarks_dir else None
                 with self._recorded_timings.record(k):
                     fpm.initialize()
-                    fpm.run(func, remarks_file=remarks_file)
+                    fpm.run(func)
                     fpm.finalize()

     def _optimize_final_module(self):
@@ -682,9 +678,13 @@ class CPUCodeLibrary(CodeLibrary):
         if not config.LLVM_REFPRUNE_PASS:
             self._final_module = remove_redundant_nrt_refct(self._final_module)
         full_name = "Module passes (full optimization)"
+        remarks_dir = os.environ.get("NUMBA_LLVM_OPT_REMARKS_DIR", None)
+        remarks_file = os.path.join(remarks_dir,
+                                    "%s-remarks.yaml" % self.name
+                                    ) if remarks_dir else None
         with self._recorded_timings.record(full_name):
             # The full optimisation suite is then run on the refop pruned IR
-            self._codegen._mpm_full.run(self._final_module)
+            self._codegen._mpm_full.run(self._final_module, remarks_file=remarks_file)

     def _get_module_for_linking(self):
         """
```

Reply:
with that, i get:

```
--- !Analysis
Pass:            loop-vectorize
Name:            CFGNotUnderstood
DebugLoc:        { File: optremarks.py, Line: 7, Column: 1 }
Function:        _ZN8__main__3fooB2v1B54c8tJTC_2fWwKbcFpQAdBTgEQrlo6UyVFGMNAWVBgYHmKUJAA_3d_3dE5ArrayIfLi1E1C7mutable7alignedE
Args:
  - String:          'loop not vectorized: '
  - String:          loop control flow is not understood by vectorizer
...
--- !Analysis
Pass:            loop-vectorize
Name:            CFGNotUnderstood
DebugLoc:        { File: optremarks.py, Line: 7, Column: 1 }
Function:        _ZN8__main__3fooB2v1B54c8tJTC_2fWwKbcFpQAdBTgEQrlo6UyVFGMNAWVBgYHmKUJAA_3d_3dE5ArrayIfLi1E1C7mutable7alignedE
Args:
  - String:          'loop not vectorized: '
  - String:          loop control flow is not understood by vectorizer
...
```
from source:

```python
from numba import njit
import numpy as np


@njit(debug=True)
def foo(arr):
    c = arr + arr.sum()
    return c

res = foo(np.arange(100).astype(np.float32))
print(res)
```